File: BlokzincirAnalizi/BlokzincirAnalizi/MerkleAgaci.cs
using System.Security.Cryptography;
using System.Text;

namespace BlokzincirAnalizi
{
    /// <summary>
    /// SHA-256 tabanlı Merkle Ağacı.
    ///
    /// Blokzincirde bir işlem grubunun bütünlüğünü tek bir
    /// "Merkle Root" hash değeriyle özetlemek için kullanılır.
    /// Herhangi bir işlem değiştirildiğinde kök hash tamamen değişir.
    ///
    /// Algoritma (özyinelemeli):
    ///   1. Her tx_id'yi SHA-256 ile yaprak hash'e dönüştür.
    ///   2. Çift sayı yoksa son yaprağı kopyala (Bitcoin standardı).
    ///   3. Her komşu çifti birleştirip hash'le → üst seviye.
    ///   4. Tek hash kalana dek tekrarla → Merkle Root.
    /// </summary>
    public class MerkleAgaci
    {
        /// <summary>Nihai Merkle Root hash değeri.</summary>
        public string Kok { get; private set; } = string.Empty;

        /// <summary>Yaprak düğümleri: her tx_id'nin SHA-256 hash'i.</summary>
        public List<string> Yapraklar { get; private set; } = [];

        /// <summary>Her seviyenin hash listesini saklar (görselleştirme için).</summary>
        public List<List<string>> Seviyeler { get; private set; } = [];

        private readonly List<Islem> _islemler;

        /// <param name="islemler">Merkle ağacına dahil edilecek işlem listesi</param>
        public MerkleAgaci(List<Islem> islemler)
        {
            _islemler = islemler;
            Yapraklar = islemler.Select(i => Sha256(i.TxId)).ToList();
            Kok = Insa(new List<string>(Yapraklar));
        }

        // ---- Yardımcı Metotlar ----

        /// <summary>Verilen string'i UTF-8 kodlayıp SHA-256 ile hash'ler.</summary>
        private static string Sha256(string veri)
        {
            using var sha = SHA256.Create();
            byte[] hash = sha.ComputeHash(Encoding.UTF8.GetBytes(veri));
            return Convert.ToHexString(hash).ToLower();
        }

        /// <summary>
        /// Özyinelemeli Merkle ağacı inşası.
        /// Her çağrıda mevcut seviyeyi Seviyeler listesine kaydeder.
        /// </summary>
        private string Insa(List<string> dugumler)
        {
            // Boş işlem listesi durumu
            if (dugumler.Count == 0)
                return Sha256("BOS_AGAC");

            // Mevcut seviyeyi kaydet
            Seviyeler.Add(new List<string>(dugumler));

            // Tek düğüm kaldı → bu düğüm Merkle Root'tur
            if (dugumler.Count == 1)
                return dugumler[0];

            // Tek sayıda düğüm varsa son düğümü kopyala
            if (dugumler.Count % 2 != 0)
                dugumler.Add(dugumler[^1]);

            // İkişerli çiftleri birleştirip hash'le → üst seviye
            var ustSeviye = new List<string>();
            for (int i = 0; i < dugumler.Count; i += 2)
                ustSeviye.Add(Sha256(dugumler[i] + dugumler[i + 1]));

            // Özyinelemeli çağrı: bir üst seviyeyle devam et
            return Insa(ustSeviye);
        }

        // ---- Genel Metotlar ----

        /// <summary>
        /// Verilen tx_id'nin bu Merkle ağacında kayıtlı olup olmadığını doğrular.
        /// tx_id'yi SHA-256 ile hash'leyerek yaprak listesinde arar.
        /// </summary>
        public bool DogrulaIslem(string txId)
        {
            string hash = Sha256(txId);
            return Yapraklar.Contains(hash);
        }

        /// <summary>Ağacın seviye sayısını (derinliğini) döndürür.</summary>
        public int Derinlik => Seviyeler.Count;

        /// <summary>Ağaçtaki toplam işlem sayısı.</summary>
        public int IslemSayisi => _islemler.Count;

        public override string ToString() =>
            $"MerkleAgaci(işlem={IslemSayisi}, derinlik={Derinlik}, kök={Kok[..12]}...)";
    }
}

File: merkle_agaci.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
merkle_agaci.py
---------------
SHA-256 tabanlı Merkle Ağacı implementasyonu.

Merkle Ağacı, blokzincirde bir işlem grubunun bütünlüğünü
tek bir kök hash (Merkle Root) ile özetlemek için kullanılır.
Herhangi bir işlem değiştirildiğinde kök hash tamamen değişir.

Bağımlılık: islem.py → Transaction
"""

from __future__ import annotations
import hashlib
from islem import Transaction


class MerkleTree:
    """
    SHA-256 tabanlı Merkle Ağacı.

    Yapı
    ----
    - Yapraklar (leaves) : Her Transaction.tx_id'nin SHA-256 hash'i
    - Ara seviyeler      : Komşu çift hash'lerin birleştirilip tekrar hash'lenmesi
    - Kök (root)         : Tüm ağacı özetleyen tek SHA-256 hash değeri

    Algoritma (özyinelemeli)
    ------------------------
    1. tx_id listesini SHA-256 ile yaprak listesine dönüştür.
    2. Çift sayı yoksa son yaprağı kopyala (Bitcoin standardı).
    3. Her çifti birleştirip hash'le → bir üst seviye oluştur.
    4. Tek hash kalana dek tekrarla → Merkle Root.

    Bağımlılık: islem.Transaction
    """

    def __init__(self, transactions: list[Transaction]) -> None:
        self.transactions: list[Transaction] = transactions

        # Yaprak düğümleri: her tx_id'nin SHA-256 hash'i
        self.leaves: list[str] = [
            self._sha256(tx.tx_id) for tx in transactions
        ]

        # Her seviyenin hash listesini saklar (görselleştirme için)
        self.levels: list[list[str]] = []

        # Ağacı inşa et ve kökü hesapla
        self.root: str = self._build(self.leaves.copy())

    # ---- Özel Metotlar ----

    @staticmethod
    def _sha256(data: str) -> str:
        """Verilen string'i UTF-8 kodlayıp SHA-256 ile hash'ler."""
        return hashlib.sha256(data.encode("utf-8")).hexdigest()

    def _build(self, nodes: list[str]) -> str:
        """
        Özyinelemeli Merkle ağacı inşası.

        Her çağrıda mevcut seviyeyi self.levels'a kaydeder,
        bir üst seviyeyi hesaplayarak kendini tekrar çağırır.
        Tek düğüm kalana dek devam eder.

        Parameters
        ----------
        nodes : Mevcut seviyedeki hash listesi

        Returns
        -------
        Merkle Root hash string'i
        """
        # Boş işlem listesi durumu
        if not nodes:
            return self._sha256("EMPTY_TREE")

        # Mevcut seviyeyi kaydet (görselleştirme için)
        self.levels.append(nodes.copy())

        # Tek düğüm kaldı → bu düğüm Merkle Root'tur
        if len(nodes) == 1:
            return nodes[0]

        # Tek sayıda düğüm varsa son düğümü kopyala
        if len(nodes) % 2 != 0:
            nodes.append(nodes[-1])

        # İkişerli çiftleri hash'leyerek bir üst seviye oluştur
        parent_level: list[str] = []
        for i in range(0, len(nodes), 2):
            combined = nodes[i] + nodes[i + 1]
            parent_level.append(self._sha256(combined))

        # Özyinelemeli çağrı: bir üst seviyeyle devam et
        return self._build(parent_level)

    # ---- Genel Metotlar ----

    def verify_transaction(self, tx_id: str) -> bool:
        """
        Verilen tx_id'nin bu Merkle ağacına ait olup olmadığını doğrular.

        tx_id'yi SHA-256 ile hash'leyerek yaprak listesinde arar.

        Parameters
        ----------
        tx_id : Doğrulanmak istenen işlem kimliği

        Returns
        -------
        True  → işlem Merkle ağacında kayıtlı
        False → işlem bulunamadı veya değiştirilmiş
        """
        candidate_hash = self._sha256(tx_id)
        return candidate_hash in self.leaves

    def depth(self) -> int:
        """Ağacın seviye sayısını (derinliğini) döndürür."""
        return len(self.levels)

    def __repr__(self) -> str:
        return (
            f"MerkleTree(transactions={len(self.transactions)}, "
            f"depth={self.depth()}, root={self.root[:12]!r}…)"
        )
